SLSA v1.0: overhaul requirements by party. (#536)
slsa-framework/slsa: docs/spec/v1.0/requirements.md

    provenance SHOULD be the same at all Build levels, but a few fields MAY be
    absent at L1 if they are prohibitively expensive to implement.
-   *Authenticity:* No requirements.
-   *Accuracy:* No requirements.

**TODO:** Link to local copy of provenance (../../provenance/v1.0.md)
once [#525](https://github.com/slsa-framework/slsa/pull/525) is merged.

[SLSA Provenance]: https://deploy-preview-525--slsa.netlify.app/provenance/v1.0

<td>✓<td>✓<td>✓
<tr id="provenance-authenticated"><td>Provenance Authenticated<td>

*Authenticity:* Consumers MUST be able to authenticate the provenance such that
they can:

-   Identify the transitive closure of the build system, i.e. which entities
    they must trust.
-   Verify the integrity of the provenance, i.e. that it was not tampered with
    after the build.

This SHOULD be through a digital signature from a private key accessible only to
the service generating the provenance.

*Accuracy:* The provenance MUST be generated by the build system (i.e. within
the trust boundary identified in the provenance) and not by a tenant of the
build system (i.e. outside the trust boundary).

> -   The data in the provenance MUST be obtained from the build service, either
>     because the generator *is* the build service or because the provenance
>     generator reads the data directly from the build service.

Reviewer: ```suggestion
-   The data in the provenance MUST be obtained from the build service, either
    because the provenance generator *is* the build service or because the
    provenance generator reads the data directly from the build service.
```

<tr id="build-service">
<td>Build service
<td>

All build steps ran using some build service, not on a maintainer's
workstation.

Examples: GitHub Actions, Google Cloud Build, Travis CI.

<td> <td>✓<td>✓
<tr id="ephemeral-isolated">
<td>Ephemeral and isolated
<td>

**TODO:** Use updated wording from
[#532](https://github.com/slsa-framework/slsa/pull/532).

The build service ensured that the build steps ran in an ephemeral and isolated
environment provisioned solely for this build, free of influence from other
build instances, whether prior or concurrent.

-   It MUST NOT be possible for a build to access any secrets of the build service, such as the provenance signing key.
-   It MUST NOT be possible for two builds that overlap in time to influence one another.
-   It MUST NOT be possible for one build to persist or influence the build environment of a subsequent build.
-   Build caches, if used, MUST be purely content-addressable to prevent tampering.
-   The build SHOULD NOT call out to remote execution unless it's considered part of the "builder" within the trust boundary.
-   The build SHOULD NOT open services that allow for remote influence.

> Note: This requirement was split into "Isolated" and "Ephemeral Environment" the
> initial [draft version (v0.1)](../v0.1/requirements.md).

Reviewer: Nit: I tried to rephrase this because as written it felt awkward to read, but perhaps you intentionally had it follow the same sentence structure as other notes about requirement changes from v0.1?
```suggestion
Note: In the initial [draft version (v0.1)](../v0.1/requirements.md) this was
two separate requirements, "Isolated" and "Ephemeral Environment", which have
been combined in this version.
```

</details>

## Packaging ecosystem

[Packaging ecosystem]: #packaging-ecosystem

> > **TODO:** Is there a better term that is more obvious to most readers?

Reviewer: This is a great question. It would be good to survey a broader set of potential readers. Perhaps we can create an Issue then share the link in various OpenSSF Slack channels for wider feedback?

Reviewer: I've created https://github.com/slsa-framework/slsa/issues/538 and will share with various OpenSSF Slack channels now.

ad-hoc distribution of software, such as through a link on a website, is
considered an "ecosystem".

The packaging ecosystem is responsible for ensuring that consumers only use
artifacts ...**TODO**

**TODO:** Provenance MUST be available to the packaging ecosystem and/or
consumers. For the build system, this usually means outputting the provenance as
part of the build to allow the release process to upload it in a manner
prescribed by the packaging ecosystem.

**TODO:** Update the requirements to provide guidelines for how to implement,
showing what the options are:

-   How to define expectations: explicit vs implicit
-   Whether provenance is generated during the initial build and/or
    after-the-fact using reproducible builds
-   How provenance is distributed
-   When verification happens: during upload, during download, and/or continuous
    monitoring
-   What happens on failure: blocking, warning, and/or asynchronous notification

## Consumer

[Consumer]: #consumer

A package's <dfn>consumer</dfn> is the organization or individual that uses the
package.

The only requirement on the consumer is that they MAY have to opt-in to enable

> SLSA verification, depending on the packaging ecosystem.

Reviewer: Is it worth reinforcing the 'verify artifacts' principal here?

```suggestion
SLSA verification, depending on the packaging ecosystem. All packages with SLSA
provenance SHOULD have their provenance verified at use, in an ideal scenario
that verification is performed automatically by the package ecosystem's client.
```

> **TODO:** Anything else? Do they need to make risk-based decisions? Respond to

> > errors/warnings?

Reviewer: Yes. Depending on ecosystem the consumer will have to make some choices, so it's worth highlighting that here.
